include/zephyr/zbus/zbus.h

```diff
@@ -0,0 +1,411 @@
+			ZBUS_USER_DATA_INIT ZBUS_RUNTIME_OBSERVERS_LIST_INIT.message_size =        \
+				sizeof(_type),                                                     \
+			.message = &_CONCAT(_zbus_message_, _name), .validator = _validator,       \
+			.mutex = &_CONCAT(_zbus_mutex_, _name),                                    \
+			.observers = (struct zbus_observer **)(struct zbus_observer *[]){          \
+				FOR_EACH_NONEMPTY_TERM(ZBUS_REF, (,), _observers) NULL}}
+
+/**
+ * @brief Initialize a message.
+ *
+ * This macro initializes a message by passing the values to initialize the message struct
+ * or union.
+ *
+ * @param[in] _val Variadic with the initial values. ``ZBUS_INIT(0)`` means ``{0}``, as
+ * ZBUS_INIT(.a=10, .b=30) means ``{.a=10, .b=30}``.
+ */
+#define ZBUS_MSG_INIT(_val, ...)                                                                   \
+	{                                                                                          \
+		_val, ##__VA_ARGS__                                                                \
+	}
+
+/**
+ * @brief Define and initialize a subscriber.
+ *
+ * This macro defines an observer of subscriber type. It defines a message queue where the
+ * subscriber will receive the notification asynchronously, and initialize the ``struct
+ * zbus_observer`` defining the subscriber.
+ *
+ * @param[in] _name The subscriber's name.
+ * @param[in] _queue_size The notification queue's size.
+ */
+#define ZBUS_SUBSCRIBER_DEFINE(_name, _queue_size)                                                 \
+	K_MSGQ_DEFINE(_zbus_observer_queue_##_name, sizeof(struct zbus_channel *), _queue_size,    \
+		      sizeof(struct zbus_channel *));                                              \
+	_ZBUS_STRUCT_DECLARE(zbus_observer, _name) = {                                             \
+		.name = STRINGIFY(_name), .enabled = true, .queue = &_zbus_observer_queue_##_name, \
+				  .callback = NULL}
+
+/**
+ * @brief Define and initialize a listener.
+ *
+ * This macro defines an observer of listener type. This macro establishes the callback where the
+ * listener will be notified synchronously, and initialize the ``struct zbus_observer`` defining the
+ * listener.
+ *
+ * @param[in] _name The listener's name.
+ * @param[in] _cb The callback function.
+ */
+#define ZBUS_LISTENER_DEFINE(_name, _cb)                                                           \
+	_ZBUS_STRUCT_DECLARE(zbus_observer, _name) = {                                             \
+		.name = STRINGIFY(_name), .enabled = true, .queue = NULL, .callback = (_cb)}
+
+/**
+ *
+ * @brief Publish to a channel
+ *
+ * This routine publishes a message to a channel.
+ *
+ * @param chan The channel's reference.
+ * @param msg Reference to the message where the publish function copies the channel's
+ * message data from.
+ * @param timeout Waiting period to publish the channel,
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
```

Comment:
The `timeout` parameter is a bit misleading. It's currently used to lock the channel's mutex, but doesn't take the observer notify timeouts into account. The total call time can be greater than the timeout and still be successful.